### ISSUE TITLE
last_submission

### ISSUE THREAD

**Issue by tigrs50** [tigrs50/tigrs50#1]
# トレーディング環境
class TradingEnvWithSineThreshold(gym.Env):
    def __init__(self, data, steps_ahead_min=2, steps_ahead_max=7, max_buy_fraction=0.5, max_sell_fraction=0.5, threshold=0.05, observation_period=12, render_mode=None):
        super(TradingEnvWithSineThreshold, self).__init__()

        self.open_col = 'Open'
        self.high_col = 'High'
        self.low_col = 'Low'
        self.close_col = 'Close'
        self.volume_col = 'Volume'

        # アクションスペース: 取引アクションの大きさ（0~1）とsteps_ahead（-1~1）
        self.action_space = gym.spaces.Box(low=np.array([0, -1]),
                                           high=np.array([1, 1]),
                                           dtype=np.float32)

        # 観測スペース: 価格変化率 + Volume
        self.observation_space = gym.spaces.Box(low=-np.inf, high=np.inf, shape=(5,), dtype=np.float32)

        self.data = data
        self.current_step = 0  # 現在のステップ
        self.cash = 10000  # 初期キャッシュ
        self.stock_held = 0  # 保有株数
        self.initial_price = self.data[self.open_col].iloc[0]  # 初期価格を取得

        # steps_ahead の最小値と最大値を設定
        self.steps_ahead_min = steps_ahead_min
        self.steps_ahead_max = steps_ahead_max

        # 最大取引量の制限
        self.max_buy_fraction = max_buy_fraction  # 最大購入割合（例: 0.5 = 資金の50%）
        self.max_sell_fraction = max_sell_fraction  # 最大売却割合（例: 0.5 = 保有株数の50%）

        # 取引を行うためのしきい値（target_magnitude * sin(steps_ahead_angle)がこれを超える必要がある）
        self.threshold = threshold

        self.observation_period = observation_period  # 初期の観測期間
        self.rewards = []
        self.cash_history = []
        self.stock_history = []

        self.render_mode = render_mode

    def reset(self):
        """エピソードをリセット"""
        self.current_step = 0
        self.cash = 10000
        self.stock_held = 0
        self.initial_price = self.data[self.open_col].iloc[0]
        self.rewards = []
        self.cash_history = []
        self.stock_history = []
        return self._next_observation()

    def _next_observation(self):
        """次の観測データを取得"""
        open_price = self.data[self.open_col].iloc[self.current_step]
        current_price = self.data[self.close_col].iloc[self.current_step]
        high_price = self.data[self.high_col].iloc[self.current_step]
        low_price = self.data[self.low_col].iloc[self.current_step]
        volume = self.data[self.volume_col].iloc[self.current_step]

        # 各価格に対する初期価格からの変化率を計算
        open_percentage_change = ((open_price - self.initial_price) / self.initial_price) * 100
        percentage_change = ((current_price - self.initial_price) / self.initial_price) * 100
        high_percentage_change = ((high_price - self.initial_price) / self.initial_price) * 100
        low_percentage_change = ((low_price - self.initial_price) / self.initial_price) * 100

        # 変化率と取引量を観測として返す
        return np.array([open_percentage_change, percentage_change, high_percentage_change, low_percentage_change, volume])

    def step(self, action):
        """環境の1ステップを実行"""
        target_magnitude, steps_ahead_angle = action

        # steps_ahead_angle (-1から1) を -90度から90度の角度にマッピング
        angle = steps_ahead_angle * 90  # -1~1 を -90~90度に変換
        # 角度を steps_ahead_min から steps_ahead_max に変換
        steps_ahead = self.steps_ahead_min + (np.abs(angle) / 90) * (self.steps_ahead_max - self.steps_ahead_min)

        # target_magnitude * sin(角度に変換したsteps_ahead) を計算
        sin_angle = np.sin(np.radians(angle))
        dynamic_threshold_value = target_magnitude * sin_angle

        # しきい値（0.3）を超えているか確認
        if dynamic_threshold_value >= self.threshold:
            current_price = self.data[self.close_col].iloc[self.current_step]

            # 取引アクションの処理: 正のtarget_magnitudeは買い、負のtarget_magnitudeは売り
            if target_magnitude > 0:  # 買いアクション
                # 最大購入量に制限をかける（self.cash の max_buy_fraction の範囲内で購入）
                buy_amount = min(target_magnitude * self.cash, self.cash * self.max_buy_fraction)
                num_shares = buy_amount // current_price
                self.stock_held += num_shares
                self.cash -= num_shares * current_price

            elif target_magnitude < 0:  # 売りアクション
                # 最大売却量に制限をかける（self.stock_held の max_sell_fraction の範囲内で売却）
                sell_amount = min(-target_magnitude * self.stock_held, self.stock_held * self.max_sell_fraction)
                self.cash += sell_amount * current_price
                self.stock_held -= sell_amount

        # steps_ahead に基づいて次のステップに進む
        self.current_step = min(self.current_step + int(steps_ahead), len(self.data) - 1)  # 範囲を超えないよう修正

        # 報酬を計算
        reward = self.cash + self.stock_held * self.data[self.close_col].iloc[self.current_step] - 10000
        self.rewards.append(reward)
        self.cash_history.append(self.cash)
        self.stock_history.append(self.stock_held)

        # エピソードが終了したかを判定
        done = self.current_step >= len(self.data) - 1
        obs = self._next_observation()

        return obs, reward, done, {}

    def render(self, mode='human'):
        """環境の状態を出力"""
        if self.render_mode == 'human':
            print(f"Step: {self.current_step}, Cash: {self.cash}, Stock Held: {self.stock_held}, Price: {self.data[self.close_col].iloc[self.current_step]}")

# モデルの学習を実行
def run_sac_with_sine_threshold(data, learning_rate=3e-4, num_stocks=None):
    # データから総銘柄数を計算 (6列ごとに1銘柄)
    total_stocks = int(data.shape[1] / 6)
    if num_stocks is None or num_stocks > total_stocks:
        num_stocks = total_stocks

    print(f"学習する銘柄の数: {num_stocks} / {total_stocks}")

    model = None

    for i in tqdm(range(num_stocks), desc="Training progress"):
        tqdm.write(f"現在の銘柄: 銘柄{i+1}")

        # 各銘柄のデータを取得 (6列: Open, High, Low, Close, Adj Close, Volume)
        stock_data = data.iloc[:, i * 6:(i + 1) * 6]
        stock_data.columns = ['Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume']

        # NaNが含まれている銘柄はスキップ
        if stock_data.isna().sum().sum() > 0:
            tqdm.write(f"銘柄{i+1} のデータにNaNが含まれているためスキップします。")
            continue

        # 環境の作成 (sinベースのしきい値とsteps_aheadの最適化を含む)
        env = DummyVecEnv([lambda: TradingEnvWithSineThreshold(
            stock_data, steps_ahead_min=2, steps_ahead_max=7, max_buy_fraction=0.5, max_sell_fraction=0.5, threshold=0.05, observation_period=min(12, len(stock_data)))])

        if model is None:
            # モデルの初期化
            model = SAC("MlpPolicy", env, verbose=1, learning_rate=learning_rate, target_update_interval=2, buffer_size=10000000)
        else:
            # モデルの環境設定を変更
            model.set_env(env)

        # 銘柄ごとのタイムステップ数を取得
        total_timesteps = len(stock_data)
        tqdm.write(f"銘柄{i+1} の行数: {total_timesteps}")

        # モデルの学習
        model.learn(total_timesteps=total_timesteps)

    # 学習後のモデルを保存
    model.save("sac_trading_model")

if __name__ == "__main__":
    # データファイルパスを指定 (ユーザーは適切なパスを設定する必要があります)
    file_path = '/content/aaaa.csv'  # ファイルパスを適切に設定
    data_combined = load_data(file_path)

    num_stocks_to_train = 2000  # 学習する銘柄数を指定

    # 指定した銘柄数で学習を実行
    run_sac_with_sine_threshold(data_combined, num_stocks=num_stocks_to_train)
